### PR TITLE
[ESP32] Options to include Device instance information to factory partition

### PR DESCRIPTION
**Pull request by shubhamdp**
#### Problem
Fixes #19996

#### Change overview
Added options to include device instance information into factory partition.

#### Testing
Manually tested: generated the factory partition with using script and verified by reading basic cluster attributes

### CODE REVIEW

File: scripts/tools/generate_esp32_chip_factory_bin.py [hunk 1]
```diff
@@ -41,24 +41,142 @@
...
+
 def check_tools_exists():
     TOOLS['spake2p'] = shutil.which('spake2p')
     if TOOLS['spake2p'] is None:
         logging.error('spake2p not found, please add spake2p path to PATH environment variable')
         sys.exit(1)
 
 
+def check_str_range(s, min_len, max_len, name):
+    if s and (len(s) not in range(min_len, max_len + 1)):
```
Comment (tcarmelveilleux) [inline scripts/tools/generate_esp32_chip_factory_bin.py:151]:
```suggestion
    if s and ((len(s) < min_len) or (len(s) > max_len)):
```

File: scripts/tools/generate_esp32_chip_factory_bin.py [hunk 2]
```diff
@@ -41,24 +41,142 @@
...
+        logging.error('%s must be between %d and %d characters', name, min_len, max_len)
+        sys.exit(1)
+
+
+def check_int_range(value, min_value, max_value, name):
+    if value and (value not in range(min_value, max_value + 1)):
```
Comment (tcarmelveilleux) [inline scripts/tools/generate_esp32_chip_factory_bin.py:157]:
```suggestion
    if value and ((value < min_value) or (value > max_value)):
```

File: scripts/tools/generate_esp32_chip_factory_bin.py [hunk 3]
```diff
@@ -164,15 +308,31 @@ def any_base_int(s): return int(s, 0)
...
     parser.add_argument('-s', '--size', type=any_base_int, required=False, default=0x6000,
                         help='The size of the partition.bin, default: 0x6000')
 
     args = parser.parse_args()
     validate_args(args)
     check_tools_exists()
     spake2p_params = gen_spake2p_params(args.passcode)
-    generate_nvs_bin(args, spake2p_params)
+    populate_factory_data(args, spake2p_params)
+    gen_raw_ec_keypair_from_der(args.dac_key, FACTORY_DATA['dac-pub-key']['value'], FACTORY_DATA['dac-key']['value'])
```
Comment (tcarmelveilleux) [inline scripts/tools/generate_esp32_chip_factory_bin.py:332]: Is it possible to encrypt the DAC key using a device-specific key, or is this already done by the NVS layer, or does the device attestation private key appear directly in the flash?

**Review by tcarmelveilleux — APPROVED** [review]

**Review by dhrishi — APPROVED** [review]

**Review by tehampson — APPROVED** [review]

File: scripts/tools/generate_esp32_chip_factory_bin.py [hunk 4]
```diff
@@ -164,15 +308,31 @@ def any_base_int(s): return int(s, 0)
...
+    parser.add_argument('--vendor-id', type=any_base_int, required=False, help='Vendor id')
+    parser.add_argument('--vendor-name', type=str, required=False, help='Vendor name')
+    parser.add_argument('--product-id', type=any_base_int, required=False, help='Product id')
+    parser.add_argument('--product-name', type=str, required=False, help='Product name')
+    parser.add_argument('--hw-ver', type=any_base_int, required=False, help='Hardware version')
+    parser.add_argument('--hw-ver-str', type=str, required=False, help='Hardware version string')
+    parser.add_argument('--mfg-date', type=str, required=False, help='Manufacturing date in format YYYY-MM-DD')
+    parser.add_argument('--serial-num', type=str, required=False, help='Serial number')
+    parser.add_argument('--unique-id', type=str, required=False,
+                        help='128-bt unique identifier, provide 32-byte hex string, e.g. "1234567890abcdef1234567890abcdef"')
```
[the remark this answers is not in the document]
Comment (tehampson) [reply scripts/tools/generate_esp32_chip_factory_bin.py]: Ops... wow I should have put that together. Thanks

**Pull request merged**